rfc/accepted/0021-command-driven-connectors.md

```diff
@@ -0,0 +1,126 @@
+  end;
+  
+  create connector read_file_list from read_file_list;
+  create connector file_connector from file_connector;
+  create connector file_list from file_list;
+  
+  connect /connector/read_file_list/out to /connector/file_list/control;
+  connect /connector/file_list/out to /pipeline/main/in;
+  
+  # This is the magic - we send the commands here, note the "control" port
+  connect /pipeline/main/out to /connector/file_connector/control;
+  
+  connect /connector/file_connector/out to /pipeline/main/out;
+end;
+```
+
+## Reference-level Explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Each command driven connector implements at least one port - `control`.
+`control` is an input port, through which the commands are sent.
+
+The commands are standardised as far as it is practical, so the connectors can be swapped without adjusting the rest of the system.
+Convenience functions are provided to generate the events.
+Each event has a `correlation` field, which will be a unique identifier provided in the metadata for the response. This is automatically generated by the convenience functions.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+This raises the complexity of Tremor.
+
+## Rationale and Alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+There are no known alternatives that provide the same benefits.
+Currently, for example the S3 reader connector, will read all files in the bucket, once, which has limited use.
+
+## Prior Art
+[prior-art]: #prior-art
+
+Discuss prior art, both the good and the bad, in relation to this proposal.
+A few examples of what this can include are:
+
+- For language, library, tools, and clustering proposals: Does this feature exist in other programming languages, and what experience have their community had?
+- For community proposals: Is this done by some other community and what were their experiences with it?
+- For other teams: What lessons can we learn from what other communities have done here?
+- Papers: Are there any published papers or great posts that discuss this? If you have some relevant papers to refer to, this can serve as a more detailed theoretical background.
+
+This section is intended to encourage you as an author to think about the lessons from other projects, provide readers of your RFC with a fuller picture.
+If there is no prior art, that is fine- your ideas are interesting to us whether they are brand new or if it is an adaptation from other projects.
+
+Note that while precedent set by other projects is some motivation, it does not, on its own, motivate an RFC.
+Please also take into consideration that Tremor sometimes intentionally diverges from similar projects.
+
+## Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- How do we enforce uniformity of commands across connectors?
+- How would writes work?
+- Are multiple events per command allowed in the output?

```

Comment:
My gut feeling is that batching for commands is possible, we can unbatch them and go about our work, I don't see anything that would prevent this. On the other side I don't see much of a use for this either.

Reply:
We probably want commands from a given command source to be applied in arrival order - that would seem to be a strong enough guarantee initially and consistent with what we already have for user or runtime events. Later, if we need stronger or variant guarantees - perhaps these could be controlled via command traits. A future worth iterating in this RFC IMO ...